### PR TITLE
Made ByteSynkArray to use shared_ptr

### PR DESCRIPTION
**Pull request by Trial97**
ok this should fix https://discord.com/channels/1031648380885147709/1032130809055940690/threads/1118963246330810469 and maybe https://github.com/PrismLauncher/PrismLauncher/issues/1199

### CODE REVIEW

**Comment by Ryex** [comment]
when I attempted this change I ended up with leaked shared pointers because of cyclic references which is why it wasn't part of the initial memory leak fixes. Have you checked with address sanitizer that these don't leak?

**Comment by Trial97** [comment]
Yeah,leaks:
```
SUMMARY: AddressSanitizer: 1506594 byte(s) leaked in 6322 allocation(s).
```
Let me tweak more.

**Comment by Trial97** [comment]
@Ryex can you help me with the address sanitizer?
I added the flags for the build: `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address -fno-omit-frame-pointer")` and added this env var: `export ASAN_OPTIONS=fast_unwind_on_malloc=0:detect_stack_use_after_return=1:check_initialization_order=1:strict_init_order=1:malloc_context_size=100 `
Now the question is how do I determine if I have a problem with the shared_ptr?
My initial test case was:
1. open the generated executable
2. add a new instance
3. select all the mods8 (first mod pack from curse forge)
4. go to the process of adding the mod pack
5. after all is done just close the prism(completely)
6. on close I was greeted with some leaks:`Indirect leak of 34 byte(s) in 3 object(s) allocated from:`

My problem is if I apply the same to develop I get similar results.
Should I open an issue with this?
*Also this is not limited to curseforge mod packs: except modrinth, all providers have similar results for the analyzer(same steps)

**Comment by Ryex** [comment]
if your leaking the shared pointer you should get stack traces leading under `make_shared` when you exit

shared pointers leak when there is memory behind a shared pointer which holds a shared pointer that either directly or indirectly holds a shared pointer to the parent. In theory the solution is using weak pointers at some point in that chain to break the cycle.

I really did try my best to remove all the leaks when I made that PR but it could be possible new ones were introduced or that I missed some.

---

The trick to working with the address sanitizer, especially with indirect leaks is to see where the memory that leaks was initialized and try to trace the pointer mack to it's nearest owner. then go through and try to find where it should be drooped. if that pointer traces back to the ownership of a shared pointer try to find the cycle.

the solution may be that we need to start using weak pointers for instances passes to dialogs or pages out of their ownership in the Application class.

---

The other place to be extra careful with shared pointers is anywhere they may be captured by a lambda that is connected to a signal or slot. if you can, only capture the raw pointer using `get` or explicitly capture by reference so that a copy isn't made increasing the ref-count.

**Comment by Trial97** [comment]
Ok so I give up regarding the address sanitizer thing.
I tested: all resource download, no leak there; all modpacks download, except modrinth all the other providers has some leak; also tested the mod update, and found leaks there too.
But no leaks related to `std::make_shared<QByteArray>` or `std::shared_ptr<QByteArray>` and all the leaks I was able to reproduce them on develop.
***
Also for me, the following character creates a leak: `✔`(https://github.com/PrismLauncher/PrismLauncher/blob/fd9a8d1551e5736aff6da10d8d00cc631c6d6ee0/launcher/ui/dialogs/BlockedModsDialog.cpp#L166) removing it the leak is fixed :exploding_head:

**Comment by Trial97** [comment]
Ok status update: well fixed the found leaks in modpacks download(If needed I can create a separate PR with only them)
Except the one caused by the character mention previously.
Did not find any leak to lead to the ByteSyncArray, or associated job. (tested on all resource downloads)

**Review by Ryex — APPROVED** [review]
LGTM

File: launcher/modplatform/helpers/HashUtils.h [hunk 1]
```diff
@@ -21,6 +22,9 @@ class Hasher : public Task {
     QString getResult() const { return m_hash; };
     QString getPath() const { return m_path; };
 
+   signals:
+    void getResults(QString hash);
```
Comment (TheKodeToad) [inline launcher/modplatform/helpers/HashUtils.h]: Maybe `resultsReady` would be a better name? This looks like a getter

File: launcher/modplatform/atlauncher/ATLPackInstallTask.cpp [hunk 2]
```diff
@@ -99,11 +99,12 @@ void PackInstallTask::onDownloadSucceeded()
     qDebug() << "PackInstallTask::onDownloadSucceeded: " << QThread::currentThreadId();
     jobPtr.reset();
 
-    QJsonParseError parse_error {};
-    QJsonDocument doc = QJsonDocument::fromJson(response, &parse_error);
-    if(parse_error.error != QJsonParseError::NoError) {
-        qWarning() << "Error while parsing JSON response from ATLauncher at " << parse_error.offset << " reason: " << parse_error.errorString();
-        qWarning() << response;
+    QJsonParseError parse_error{};
+    QJsonDocument doc = QJsonDocument::fromJson(*response.get(), &parse_error);
```
Comment (TheKodeToad) [inline launcher/modplatform/atlauncher/ATLPackInstallTask.cpp]: Afaik, `.get` shouldn't be necessary here

**Review by TheKodeToad — APPROVED** [review]
\<tired review\> overall good

**Pull request merged**